### PR TITLE
[SPARK-47214][Python] Create UDTF API for 'analyze' method to differentiate constant NULL arguments and other types of arguments

### PR DESCRIPTION
**Pull request by dtenedor**
### What changes were proposed in this pull request?

This PR creates a Python UDTF API for 'analyze' method to differentiate constant NULL arguments and other types of arguments.

### Why are the changes needed?

Before this PR, each `AnalyzeArgument` had a `value` field which was non-None if the argument was a constant-foldable scalar expression:

```
@dataclass(frozen=True)
class AnalyzeArgument:
    """
    The argument for Python UDTF's analyze static method.

    Parameters
    ----------
    dataType : :class:`DataType`
        The argument's data type
    value : any, optional
        The calculated value if the argument is foldable; otherwise None
    isTable : bool
        If True, the argument is a table argument.
    """

    dataType: DataType
    value: Optional[Any]
    isTable: bool
```

However, it was also `None` if the argument is literal NULL! This is a challenge for UDTF authors who want to tell exactly which scalar arguments were literal NULL.

Now we add a new field to tell the difference:

```
isConstantExpression : bool
    If True, the argument is a constant-foldable scalar expression. Then the 'value' field
    contains None if the argument is a NULL literal, or a non-None value if the argument is a
    non-NULL literal. In this way, we can distinguish between a literal NULL argument and other
    types of arguments such as complex expression trees or table arguments where the 'value'
    field is always None.
```

### Does this PR introduce _any_ user-facing change?

Yes, see above.

### How was this patch tested?

This PR applies test coverage.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by dtenedor** [comment]
cc @dongjoon-hyun @ueshin @allisonwang-db

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/UserDefinedPythonFunction.scala [hunk 1]
```diff
@@ -207,14 +207,22 @@ class UserDefinedPythonTableFunctionAnalyzeRunner(
         case NamedArgumentExpression(k, v) => (Some(k), v)
         case _ => (None, expr)
       }
+      // Write the 'value' field to provide the argument's value if it is a foldable scalar\
```
Comment (amaliujia) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/UserDefinedPythonFunction.scala:210]: nit: extra `\` after `scalar`?

**Comment by dtenedor** [comment]
Thanks @amaliujia @ueshin for your reviews! I made the suggestions that you proposed, please take another look.

File: python/pyspark/sql/worker/analyze_udtf.py [hunk 2]
```diff
@@ -78,14 +78,16 @@ def read_arguments(infile: IO) -> Tuple[List[AnalyzeArgument], Dict[str, Analyze
...
             value = pickleSer._read_with_length(infile)
             if dt.needConversion():
                 value = dt.fromInternal(value)
         else:
             value = None
-        is_table = read_bool(infile)  # is table argument
-        argument = AnalyzeArgument(dataType=dt, value=value, isTable=is_table)
+        is_table = read_bool(infile)
+        argument = AnalyzeArgument(
+            dataType=dt, value=value, isTable=is_table, isConstantExpression=is_constant_expression)
```
Comment (ueshin) [inline python/pyspark/sql/worker/analyze_udtf.py:90]: Seems to be making the linter angry.

**Review by ueshin — APPROVED** [review]
Otherwise, LGTM.

File: python/pyspark/sql/worker/analyze_udtf.py (hunk 2 above)
Comment (dtenedor) [reply python/pyspark/sql/worker/analyze_udtf.py]: 👍 
```
spark$ dev/reformat-python 
reformatted spark/python/pyspark/sql/worker/analyze_udtf.py

All done! ✨ 🍰 ✨
1 file reformatted, 984 files left unchanged.
```

**Comment by ueshin** [comment]
The failed test seems not related to this PR.

**Comment by ueshin** [comment]
Thanks! merging to master.

**Pull request closed without merging**